Analytics recorder v5

simzzz opened the pull request.

elpiel committed:
sentry - mod analytics

simzzz committed:
v5 analytics recorder logic - still WIP

simzzz committed:
added RETURNS to db query

simzzz committed:
added missing date_trunc()

simzzz committed:
used a seperate variable for insert/update analytics query

elpiel commented on primitives/src/analytics.rs:
```diff
@@ -49,6 +50,87 @@ pub struct AnalyticsQuery {
+    }
+}
+
+impl fmt::Display for OperatingSystem {
```
You can use `parse-display` instead. It's already added to `primitives`

elpiel commented on primitives/src/sentry.rs:
```diff
@@ -136,6 +139,56 @@ pub enum Event {
+        }
+    }
+}
+
```
This should be allowed only for `postgres` features.
Do the same thing as the rest of the `From<&Row>` impls - a module

elpiel commented on sentry/migrations/20190806011140_initial-tables/up.sql:
```diff
@@ -97,3 +97,23 @@ CREATE TABLE accounting (
+
+CREATE TABLE analytics (
+    campaign_id varchar(34) NOT NULL,
+    "time" TIMESTAMP NOT NULL,
```
- [x] timestamp(2) with time zone

elpiel commented on primitives/src/ipfs.rs:
```diff
@@ -91,6 +94,29 @@ impl Url {
+    }
+
+    accepts!(TEXT, VARCHAR);
+}
```
- [x] mod postgres enabled by feature postgres

elpiel commented on primitives/src/ipfs.rs:
```diff
@@ -1,5 +1,8 @@
 
+use bytes::BytesMut;
+use postgres_types::{accepts, to_sql_checked, FromSql, IsNull, ToSql, Type};
+use std::error::Error as StdError;
```
You won't need this renaming once you move the impls to a module.

simzzz committed:
changes from PR + cleaned up some todos

simzzz committed:
removed duplicate impl

elpiel commented on sentry/src/analytics.rs:
```diff
@@ -0,0 +1,82 @@
+            country: country.to_owned(),
+            os_name: os_name.to_owned(),
+        };
+        let payout_amount = event_payout.values().sum::<Option<UnifiedNum>>().unwrap(); // TODO: Remove
```
- [x] Please move this somewhere at the beginning of the function. This will make sure that we fail fast rather than doing unnecessary work and then failing.
- [x] Include the `payout` in the `EventAnalytics` instead of sending it separately

elpiel commented on sentry/src/analytics.rs:
```diff
@@ -0,0 +1,82 @@
+    pool: &DbPool,
+    campaign: Campaign,
+    session: Session,
+    events_with_payouts: Vec<(Event, HashMap<Address, UnifiedNum>)>,
```
- [x] This should be a vec of `(Event, Address, UnifiedNum)`. We will not record the validator fees so we only need the payout to the address.

elpiel commented on sentry/src/analytics.rs:
```diff
@@ -0,0 +1,82 @@
+                    .clone() // TODO: Remove clone
+                    .ad_units
+                    .into_iter()
+                    .find(|ad_unit| ad_unit.ipfs == ipfs)
```
```suggestion
                campaign
                    .ad_units
                    .iter()
                    .find(|ad_unit| ad_unit.ipfs == ipfs)
                    .clone()
```

elpiel commented on sentry/src/analytics.rs:
```diff
@@ -0,0 +1,82 @@
+pub async fn record(
+    pool: &DbPool,
+    campaign: Campaign,
+    session: Session,
```
```suggestion
    campaign: &Campaign,
    session: &Session,
```

elpiel commented on sentry/migrations/20190806011140_initial-tables/up.sql:
```diff
@@ -97,3 +97,27 @@ CREATE TABLE accounting (
+    'Impression'
+);
+
+CREATE TABLE analytics (
```
- [x] You must update the `down.sql`

elpiel commented on sentry/src/db/analytics.rs:
```diff
@@ -222,3 +222,43 @@ pub async fn get_advanced_reports(
+    let query = "INSERT INTO events 
+    (campaign_id, time, ad_unit, ad_slot, ad_slot_type, advertiser, publisher, hostname, country, os, payout_amount, payout_count)
+    VALUES ($1, date_trunc('hour', $2), $3, $4, $5, $6, $7, $8, $9, $10, $11, 1)
+    ON CONFLICT ON CONSTRAINT channels_pkey DO UPDATE 
```
- [x] `channel_pkey` is the wrong key

elpiel commented on sentry/src/db/analytics.rs:
```diff
@@ -222,3 +222,43 @@ pub async fn get_advanced_reports(
+    VALUES ($1, date_trunc('hour', $2), $3, $4, $5, $6, $7, $8, $9, $10, $11, 1)
+    ON CONFLICT ON CONSTRAINT channels_pkey DO UPDATE 
+    SET (payout_amount = payout_amount + $11, payout_count = payout_count + 1)
+    RETURNING *";
```
- [x] specify each field, it's usually better for the query.

elpiel commented on sentry/src/db/analytics.rs:
```diff
@@ -222,3 +222,43 @@ pub async fn get_advanced_reports(
+
+    let ad_unit = Json(&event.ad_unit);
+
+    let query = "INSERT INTO events 
```
- [x] The table is not called `events`

elpiel commented on sentry/src/db/analytics.rs:
```diff
@@ -222,3 +222,43 @@ pub async fn get_advanced_reports(
     })
 }
+
+pub async fn insert_analytics(
```
- [x] Add documentation specifying that it will update a record when it's present and how it will do it (based on the `hour`). We should thing about these function names as they create a wrong impression that it will only insert new records.

elpiel commented on sentry/src/db/analytics.rs:
```diff
@@ -222,3 +222,43 @@ pub async fn get_advanced_reports(
+
+    let query = "INSERT INTO events 
+    (campaign_id, time, ad_unit, ad_slot, ad_slot_type, advertiser, publisher, hostname, country, os, payout_amount, payout_count)
+    VALUES ($1, date_trunc('hour', $2), $3, $4, $5, $6, $7, $8, $9, $10, $11, 1)
```
- [x] please use a prepared value for the `1` and bind it to a variable e.g. `initial_count` and put it on top of the function.

elpiel commented on primitives/src/ipfs.rs:
```diff
@@ -91,6 +91,36 @@ impl Url {
     }
 }
 
+mod postgres {
```
- [x] Where is the config attribute? Please check how we've done this in other modules

elpiel commented on sentry/src/db/analytics.rs:
```diff
@@ -222,3 +222,43 @@ pub async fn get_advanced_reports(
+                &event.campaign_id,
+                &event.time,
+                &ad_unit,
+                &event.ad_slot,
```
- [x] `AdUnit` implements `To` & `FromSql`, so it shouldn't require `Json()`
- [x] `AdSlot`, however, does not so please implement it.

simzzz replied:
We are getting just the IPFS for the `AdSlot` so it should be fine. I will rename the field for better clarity

elpiel commented on sentry/src/db/analytics.rs:
```diff
@@ -222,3 +222,43 @@ pub async fn get_advanced_reports(
+                &event.publisher,
+                &event.hostname,
+                &event.country,
+                &event.os_name.to_string(),
```
- [x] implement `ToSql` & `FromSql` as you will need it when fetching data from DB

elpiel commented on sentry/src/analytics.rs:
```diff
@@ -0,0 +1,82 @@
+        };
+
+        // DB: Insert or Update all events
+        let event_for_db = EventAnalytics {
```
- [x] Where is the `event_type` field? please add that too.

elpiel commented on sentry/migrations/20190806011140_initial-tables/up.sql:
```diff
@@ -97,3 +97,27 @@ CREATE TABLE accounting (
+CREATE TYPE EventType AS ENUM (
+    'Click',
+    'Impression'
+);
```
- [x] usually we use all capital letters when doing the event types. See the `AsRef<str> for Event`. However, you might just use `Varchar` for the field instead and use the value from `Event.to_string()` for the event type for now.

elpiel reviewed (changes requested):
- [x] Test the `fn record()`
- [x] Test the `insert_analytics` query.

elpiel commented on primitives/src/analytics.rs:
```diff
@@ -49,6 +50,74 @@ pub struct AnalyticsQuery {
+    Whitelisted(String),
+    #[display("Other")]
+    Other,
+}
```
Please double check if this could be done using Serde alone with `#[serde(transparent)` for the `Whitelisted` option and custom deserialize function to validate the whitelisted value when deserializing.

- [x] Write a test verifying that it can be done using only serde (e.g. use serde_json to serialize and validate the expected json)
- [x] test both Deserialize & Serialize

simzzz committed:
finished requested changes

simzzz committed:
changed analytics db so that only the ipfs of an ad unit is stored instead of the whole obj

elpiel committed:
analytics - OperatingSystem test cases

elpiel committed:
primitives - EventAnalytics - fix `From<&Row>`

elpiel commented on sentry/src/db/analytics.rs:
```diff
@@ -222,3 +222,42 @@ pub async fn get_advanced_reports(
+    VALUES ($1, date_trunc('hour', $2), $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, {})
+    ON CONFLICT ON CONSTRAINT analytics_pkey DO UPDATE 
+    SET (payout_amount = payout_amount + $12, payout_count = payout_count + 1)
+    RETURNING (campaign_id, time, ad_unit, ad_slot, ad_slot_type, advertiser, publisher, hostname, country, os, event_type, payout_amount, payout_count)", initial_count);
```
- [x] please use the prepared statement bellow for when using `initial_count`

elpiel commented on sentry/migrations/20190806011140_initial-tables/down.sql:
```diff
@@ -1,3 +1,3 @@
 -- This file should undo anything in `up.sql`
 DROP AGGREGATE jsonb_object_agg(jsonb);
-DROP TABLE event_aggregates, validator_messages, channels;
+DROP TABLE event_aggregates, validator_messages, channels, analytics;
```
Please make sure that `down.sql` works properly.
I've forgotten to add some tables for dropping (like `campaigns`) but also this would probably fail when there are records in the tables because of `ON DELETE RESTRICT`

- [x] First, drop all foreign keys
- [x] Please add all tables to `DROP TABLE`, like `campaigns`
- [x] Make sure that `down.sql` works when records are present.

elpiel reviewed (changes requested):
- [x] Test `FromSql` / `ToSql` for `OperatingSystems` - there is a bug there :wink: 
- [x] test `db::analytics::insert_analytics()`
- [x] test `analytics::record()`

simzzz committed:
some more requested changes

simzzz committed:
pushing current test progress

simzzz committed:
insert_update_and_get_analytics now works

simzzz committed:
fixed failing record() tests

simzzz committed:
cargo fmt and clippy

simzzz committed:
Merge branch 'aip-61-adex-v5' into analytics-v5

simzzz committed:
changed insert_analytics to return payout_count + changes to tests

elpiel committed:
analytics - update query params

simzzz committed:
fixed error in test

simzzz committed:
Merge branch 'analytics-v5' of https://github.com/AmbireTech/adex-validator-stack-rust into analytics-v5

simzzz committed:
Added separate Analytics structs for updating/retrieving analytics

simzzz committed:
changes to analytics - nullable slot/unit, DateHour struct, separate analytics struct for update/get, test changes

elpiel committed:
Merge branch 'aip-61-adex-v5' into analytics-v5

elpiel committed:
rustfmt

elpiel committed:
primitives - ipfs - replace imports to tokio_postgres

elpiel committed:
analytics improvements:
- primitives - sentry - DateHour
- sentry - analytics - tests improvements
- sentry - db - analytics - query & test improvements

elpiel committed:
primitives - analytics - O.S. - tests improvement
